### PR TITLE
Migrate to python-gssapi, and remove pyinstaller for RPM build

### PR DESCRIPTION
**Pull request by duncanmmacleod**
~~This PR removes the use of `pyinstaller` to bundle the requirements, instead relying on standard `Requires:` in the spec file.~~
~~This is currently failing because:~~

~~1.  `python36-kerberos` isn't available for EPEL7~~
~~1.  the EPEL9 `epel-rpm-macros` is (I think) currently broken - this will probably be resolved upstream soon enough~~

---

This PR migrates the Kerberos workflow from `pykerberos` to `gssapi`, and then removes PyInstaller from the RPM build in favour of standard `Requires` on each of the python3 packages.

This results in a working package on CentOS 7 and RockyLinux 8 (requires EPEL), but is blocked on RL9 by python3-m2crypto (see build status [here](https://koji.fedoraproject.org/koji/search?type=package&match=glob&terms=m2crypto)).

### CODE REVIEW

**Comment by duncanmmacleod** [comment]
I suppose this fairly well documents 'Python dependency hell', which is a shame.

**Comment by DrDaveD** [comment]
It turns out that I learned last week (for the vault interface in htcondor) how to handle host aliases and graceful handling of round-robin using urllib3 instead of m2crypto.  I am planning to soon update htgettoken to use the same method, which should greatly simplify things and avoid having that blocker on el9.

**Comment by duncanmmacleod** [comment]
> It turns out that I learned last week (for the vault interface in htcondor) how to handle host aliases and graceful handling of round-robin using urllib3 instead of m2crypto. I am planning to soon update htgettoken to use the same method, which should greatly simplify things and avoid having that blocker on el9.

This sounds great. Since you mentioned `urllib3`, I recommend trying to use `python-requests` as the interface, rather than going directly to `urllib3` - it has a bunch of value-added extras that are extremely useful (in my experience).

Please let me know if/when you need me to rebase/update this.

**Comment by DrDaveD** [comment]
Does python-requests have a way to override the host name expected from the host certificate?  That's needed for the htgettoken `--vaultalias` option that is sometimes useful.  I can do that with urllib3, and in the [latest update to the condor vault integration](https://github.com/htcondor/htcondor/pull/953) I use that also for controlling which IP address in a round-robin is acceptable, removing an address for 5 minutes if they have a connect time out because one of the servers is down.  The latter use case is not as important in htgettoken since it does not run for a long time, but I think it would still be useful to log which IP address times out.

**Comment by duncanmmacleod** [comment]
> Does python-requests have a way to override the host name expected from the host certificate? That's needed for the htgettoken `--vaultalias` option that is sometimes useful. I can do that with urllib3, and in the [latest update to the condor vault integration](https://github.com/htcondor/htcondor/pull/953) I use that also for controlling which IP address in a round-robin is acceptable, removing an address for 5 minutes if they have a connect time out because one of the servers is down. The latter use case is not as important in htgettoken since it does not run for a long time, but I think it would still be useful to log which IP address times out.

Honestly, I have no idea. I presume so, mainly because requests is an extremely widely used wrapper around urllib3, but I would need to study what you are currently doing to work out the requests version of the same.

**Comment by duncanmmacleod** [comment]
@DrDaveD, I have rebased this against the latest `master` - now with just three runtime dependencies (`gssapi`, `paramiko`, and `urllib3`) the RPM spec file is very simple. All of those packages are strongly supported by upstream and well maintained in in RHEL/RL (`gssapi` and `urllib3`) or EPEL (`paramiko`).

File: htgettoken.spec [hunk 1]
```diff
@@ -4,126 +4,67 @@ Summary: Get OIDC bearer tokens by interacting with Hashicorp vault
...
+
+# /usr/bin/htgettoken:
+Requires: python%{python3_pkgversion}-gssapi
+Requires: python%{python3_pkgversion}-paramiko
+Requires: python%{python3_pkgversion}-urllib3
+# /usr/bin/httokendecode:
 Requires: jq
+%if 0%{?rhel} && 0%{?rhel} >= 8
+Recommends: scitokens-cpp
+%endif
```
Comment (DrDaveD) [inline htgettoken.spec:32]: Why is rhel 7 excluded from this Recommends?  Is it not supported there yet?

**Review by DrDaveD — CHANGES_REQUESTED** [review]
Please add a commented-out Changelog entry describing this change.

I spent some time trying to get pyinstaller to work with the gssapi module, since I would rather not lock myself out of the pyinstaller option.  I made some progress, adding some `hiddenimports` for a couple of modules it failed to identify (`gssapi.raw.cython_converters` and `numbers`), but then I got stuck on a `AttributeError: mic_token` exception which I don't know what to do with.  So unfortunately it appears that the gssapi module with it's Cpython-generated modules cannot be used with pyinstaller.

**Comment by DrDaveD** [comment]
@duncanmmacleod This is waiting on you

File: htgettoken.spec (hunk 1 above)
Comment (duncanmmacleod) [reply htgettoken.spec:32]: Exactly, the old `rpmbuild` on EL7 can't handle it.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -60,18 +60,11 @@ jobs:
       - name: Install requirements
         run: |
           mamba install --name test \
-              urllib3 \
               paramiko \
+              python-gssapi \
```
[the remark this answers is not in the document]
Comment (DrDaveD) [reply .github/workflows/test.yml:64]: How confusing

**Review by DrDaveD — APPROVED** [review]

**Comment by DrDaveD** [comment]
Is there anything else you'd like to get in before I tag a version 2.0?  Like maybe fix up #40?

**Comment by DrDaveD** [comment]
Actually it would probably make sense to make the library version you were aiming for before tagging 2.0.

**Comment by duncanmmacleod** [comment]
> Actually it would probably make sense to make the library version you were aiming for before tagging 2.0.

All of my proposed mods are long term, I'm not desperate to get things I to a release. So, please release on your own schedule.

**Comment by DrDaveD** [comment]
I'm also in no hurry to make a release, and I like the idea of combining all known major changes together in one 2.0 release.

**Pull request merged**